### PR TITLE
fixes #13647 - change puppet ca proxy for several hosts at once

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

File: test/functional/hosts_controller_test.rb [hunk 1]
```diff
@@ -474,6 +474,53 @@ def setup_multiple_hosts_with_puppet
     assert_equal nil, @host2.puppet_proxy
   end
 
+  def setup_multiple_hosts_with_puppet_ca
```
Comment (unorthodoxgeek) [inline test/functional/hosts_controller_test.rb]: put the entire addition inside a describe block

``` ruby
describe "setting puppet_ca on multiple hosts" do
  before do
    #the prepare method
  end
#tests
end
```

File: test/functional/hosts_controller_test.rb [hunk 2]
```diff
@@ -474,6 +474,53 @@ def setup_multiple_hosts_with_puppet
...
+    setup_user_and_host "edit"
+    as_admin do
+      @host1, @host2 = FactoryGirl.create_list(:host, 2, :with_puppet_ca)
```
Comment (unorthodoxgeek) [inline test/functional/hosts_controller_test.rb]: @hosts = FactoryGirl.create_list

File: test/functional/hosts_controller_test.rb [hunk 3]
```diff
@@ -474,6 +474,53 @@ def setup_multiple_hosts_with_puppet
...
+    end
+  end
+
+  test "should change the puppet ca proxy of multiple hosts" do
+    @request.env['HTTP_REFERER'] = hosts_path
+    setup_multiple_hosts_with_puppet_ca
+
+    proxy = FactoryGirl.create(:smart_proxy, :features => [FactoryGirl.create(:feature, :puppetca)])
+
+    params = { :host_ids => [@host1.id, @host2.id],
```
Comment (unorthodoxgeek) [inline test/functional/hosts_controller_test.rb]: host_ids => @hosts.map(&:id)

File: test/functional/hosts_controller_test.rb [hunk 4]
```diff
@@ -474,6 +474,53 @@ def setup_multiple_hosts_with_puppet
...
+      :puppet => { :puppet_ca_proxy_id => proxy.id } }
+
+    post :update_multiple_puppet_ca_proxy, params,
+      set_session_user.merge(:user => users(:admin).id)
+
+    assert_empty flash[:error]
+
+    @host1.reload
```
Comment (unorthodoxgeek) [inline test/functional/hosts_controller_test.rb]: @hosts.each(&:reload)

or better

@hosts.each do |host|
assert_equal proxy, host.reload.puppet_ca_proxy
end

File: app/controllers/hosts_controller.rb [hunk 5]
```diff
@@ -491,6 +493,44 @@ def update_multiple_puppet_proxy
     redirect_back_or_to hosts_path
   end
 
+  def select_multiple_puppet_ca_proxy
+  end
+
+  def update_multiple_puppet_ca_proxy
```
Comment (dLobatog) [inline app/controllers/hosts_controller.rb:468]: This logic is identical to that of `update_multiple_puppet_proxy`, please refactor (extract to a method, or to a concern even) to DRY.

File: app/controllers/hosts_controller.rb [hunk 6]
```diff
@@ -869,4 +910,16 @@ def validate_multiple_puppet_proxy
       redirect_to(select_multiple_puppet_proxy_hosts_path) and return false
     end
   end
+
+  def validate_multiple_puppet_ca_proxy
```
Comment (dLobatog) [inline app/controllers/hosts_controller.rb]: Ditto, identical logic to that of `validate_multiple_puppet_proxy`, please refactor

File: app/views/hosts/select_multiple_puppet_ca_proxy.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<%= render 'selected_hosts', :hosts => @hosts %>
+
+<%= form_for :puppet, :url => update_multiple_puppet_ca_proxy_hosts_path(:host_ids => params[:host_ids]) do |f| %>
+  <%= selectable_f f,
+    :puppet_ca_proxy_id,
+    [[_("Select desired puppet ca proxy"), "disabled"]] +
```
Comment (dLobatog) [inline app/views/hosts/select_multiple_puppet_ca_proxy.html.erb]: CA (capitalized), same thing on line 7

File: app/views/hosts/select_multiple_puppet_ca_proxy.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<%= render 'selected_hosts', :hosts => @hosts %>
+
+<%= form_for :puppet, :url => update_multiple_puppet_ca_proxy_hosts_path(:host_ids => params[:host_ids]) do |f| %>
```
Comment (dLobatog) [inline app/views/hosts/select_multiple_puppet_ca_proxy.html.erb]: Nearly the same logic as `select_multiple_puppet_proxy.html.erb`, extract to a shared partial or helper

**Comment by dLobatog** [comment]
Funnily enough, yesterday I tried to delete 3 hosts from Foreman and hit this 'bug' where I cleared the Puppet proxy (using your previous PR feature) but I had to clear the CA field too.

Thanks @timogoebel , good addition, also @unorthodoxgeek for reviewing. Please fix the comments inline :smile:

**Comment by timogoebel** [comment]
@unorthodoxgeek , @dLobatog : Thanks for taking a look. I have extracted the logic into methods and refactord the tests. What do you think?

**Comment by dLobatog** [comment]
Thanks, that's better, I'll merge.

**Comment by dLobatog** [comment]
Merged as fbb47ea5535798a375a37d04254a7b5aef26b5d8, thanks @timogoebel!

**Pull request closed without merging**